### DISCUSSION TITLE
milvus2.4.9向量搜索出现随机得分为负数的情形

### DISCUSSION THREAD

**Discussion by CHENLONG871** [milvus-io/milvus#38206] [Q&A and General discussion]
很容易复现，度量方式为IP，通过设置检索参数：ef，raidus，range_filter,进行向量搜索时，如果只匹配到了一条数据，多次重复该搜索，score会随机变为负数，但是绝对值不变，这是为什么啊？

**Comment by yhmo** [comment]
IP的计算方法是 `a1*b1 + a2*b2 + ... + an*bn`
向量里如果有负值，score是可以负的

score会随机变为负数 ----------------- 多半是因为库里存在多条相同primary key的向量

**Reply by CHENLONG871** [reply]
感谢回答，1：通过给定的primary key,我这边在attu工具的 查询的时候，只能查到一条数据，所以我可以怎样直观查询到这个相同key下，在库中的不同向量数据呢？

**Reply by yhmo** [reply]
多条相同pk的数据，attu也只会显示其中一条，因为不管是query还是search，milvus server对于相同pk的结果永远只取其中一条。

**Reply by CHENLONG871** [reply]
感谢回答：那怎么理解虽然随机的出现正负，但是绝对值一样呢？

**Reply by yhmo** [reply]
IP的值只跟向量的数值相关，[1, 1]和[1, 1]的值是2，[1, 1]和[-1, -1]的值是-2。如果两条相同id的向量分别是[1, 1], [-1, -1]，那ip就是正负的

**Reply by CHENLONG871** [reply]
再次谢谢您的回复，我们的困境是，1：现在我们直观看不到同一个primary key下的多条向量数据，2：相同id的向量，可能由于我们的文本更新，产生了不同的向量，但是这里让我们很困扰的是，我这边做了什么动作，更新后的向量和之前的刚好方向相反？

**Reply by yhmo** [reply]
多少向量，多少维度，raidus，range_filter分别是多少

**Reply by CHENLONG871** [reply]
3w条向量，维度是1024, 索引类型是HNSW，metric_type:IP, raidu:0.5, range_filter:1

**Reply by yhmo** [reply]
raidu:0.5, range_filter:1
这个搜出来的值应该是0.5到1之间，也不应该有负值，你调用时具体是怎么写的入参

**Reply by CHENLONG871** [reply]
![1593ad887e4727c69e37b56727e3e11](https://github.com/user-attachments/assets/4f75ccd9-a008-4fdc-b8f7-b903568ef6ce)
![f64d3d376cb79f04d2a6fa2dcbd75b7](https://github.com/user-attachments/assets/d8f64a35-82c0-478c-b33e-499f155d2af3)

老师您看下，就是查询的参数不变，向量不变，但是score会随机的正负，出现负数的时候，好像参数没起到限制的作用

**Reply by yhmo** [reply]
Attu版本和milvus版本分别是多少？

我用Atu 2.4.8跟milvus 2.5.0，用带有负数的向量去搜，score很稳定，没有随机变化的行为

**Reply by CHENLONG871** [reply]
milvus版本2.4.11，Attu:2.4.5, 老师没有随机变化是因为这里每个primary key下只有一条数据吧，但是在我截图的这个collection下，每个primary key下按照您开头说的，不只一条向量数据；之所以不只一条向量数据的原因，很头疼，我这边不太确定这个，能确定的是，我的任务会在给定的primary key下进行过一些向量数据的更新工作，

**Reply by yhmo** [reply]
我的是有重复pk的，我是弄了3万条1024维向量，只有100个unique pk，也就是每个pk有100条重复数据，模拟下来score很稳定。
你说的向量数据的更新是指用 upsert()接口更新数据？
你分别用pymilvus这两个方法查一下行数看看：

```
from pymilvus import Collection
collection = Collection(name="xxxx")
print(collection.num_entities)

print(collection.query(expr="", output_fields=["count(*)"]))
```

**Reply by CHENLONG871** [reply]
再次谢谢答复，向量数据的更新用的是upset()接口，
![137254f0bd092ca9aaace1bac0c855b](https://github.com/user-attachments/assets/a10ba19a-6099-463b-a995-b12bddbbcfe9)

**Reply by CHENLONG871** [reply]
补充2点：1：除了upsert接口也会用到delete接口，2：attu上现在显示的数量和第二个print一致是29856
